vortex-array/src/array/sparse/compute/take.rs

```diff
@@ -1,118 +1,54 @@
-use std::convert::identity;
-
-use itertools::Itertools;
-use vortex_dtype::{match_each_integer_ptype, DType, Nullability, PType};
 use vortex_error::VortexResult;
 
-use crate::aliases::hash_map::HashMap;
-use crate::array::primitive::PrimitiveArray;
 use crate::array::sparse::SparseArray;
-use crate::array::SparseEncoding;
-use crate::compute::{take, try_cast, TakeFn};
-use crate::variants::PrimitiveArrayTrait;
-use crate::{ArrayData, IntoArrayData, IntoArrayVariant};
+use crate::array::{ConstantArray, SparseEncoding};
+use crate::compute::TakeFn;
+use crate::patches::Patches;
+use crate::{ArrayData, ArrayLen as _, IntoArrayData};
 
 impl TakeFn<SparseArray> for SparseEncoding {
     fn take(&self, array: &SparseArray, indices: &ArrayData) -> VortexResult<ArrayData> {
-        let flat_indices = indices.clone().into_primitive()?;
-        // if we are taking a lot of values we should build a hashmap
-        let (positions, physical_take_indices) = if indices.len() > 128 {
-            take_map(array, &flat_indices)?
-        } else {
-            take_search_sorted(array, &flat_indices)?
+        let patches = Patches::new(array.len(), array.indices(), array.values());
+
+        let Some(patches) = patches.take(indices)? else {
+            return Ok(ConstantArray::new(array.fill_scalar(), indices.len()).into_array());
         };
 
-        let taken_values = take(array.values(), physical_take_indices)?;
+        let (_, new_indices, new_values) = patches.into_parts();
 
-        Ok(SparseArray::try_new(
-            positions.into_array(),
-            taken_values,
+        Ok(SparseArray::try_new_with_offset(
+            new_indices,
+            new_values,
             indices.len(),
+            array.indices_offset(),
```

Comment:
I don't think this works? You can't call take on the patches that don't have an offset applied, and then still preserve the offset here

Reply:
I fixed this and added a test that would have caught the bug: `take_with_non_zero_offset`.